// uglyfly-js/index.js
/***********************************************************************
  Copyright 2014 (c) Saair Quaderi <saair.quaderi@gmail.com>

  UglyflyJS sourcecode and BSD 2-clause license info can be found here:
    https://github.com/quaderi/uglyflyjs
 ***********************************************************************/

(function () {
    'use strict';
    var Promise = require('bluebird'),
        sourceMap = require('source-map'),
        util = require('util'),
        fs = require('fs'),
        getFileContents = function getFileContents(readpath, cb) {
            fs.readFile(readpath, 'utf8', cb);
        },
        logWarning = util.error,
        logError = util.error,
        get_uglyfly_api = require('./lib/get_uglyfly_api'),
        api = get_uglyfly_api({
            Buffer: Buffer,
            Promise: Promise,
            MOZ_SourceMap: sourceMap,
            getFileContentsPromise: Promise.promisify(getFileContents),
            warn_callback: function warn_callback(txt) {
                logWarning('WARN: ' + txt);
            },
            croak_callback: function croak_callback(txt) {
                logError('ERR: ' + txt);
            }
        });
    module.exports = api;
}());
